File: KthPermutationSequence/KthPermutationSequence.cpp
// KthPermutationSequence.cpp : This file contains the 'main' function. Program execution begins and ends there.
//

#include <iostream>
#include <string>
#include <vector>

void GetPermutations(std::string& str, size_t index, std::vector<std::string>& result)
{
    if (index == str.size())
    {
        result.push_back(str);
        return;
    }

    for (size_t j = index; j < str.size(); ++j)
    {
        std::swap(str[j], str[index]);
        GetPermutations(str, index + 1, result);
        std::swap(str[j], str[index]);
    }
}

std::string GetKthPermutation(int n, int k)
{
    if (n > 62)
        throw std::runtime_error("n is too big! It should be 62 or smaller!");

    std::string str(n, '1');
    for (unsigned char i = 0; i < static_cast<unsigned char>(n); ++i)
    {
        if (i < 10)
            str[i] += i;
        else if (i < 36)
            str[i] = 'a' + i - 10;
        else
            str[i] = 'A' + i - 36;
    }

    std::vector<std::string> v;
    GetPermutations(str, 0, v);
    if (v.size() > static_cast<size_t>(k))
        return v[k - 1];

    return "";
}

int main()
{
    while (true)
    {
        std::cout << "Enter number: ";
        int n;
        std::cin >> n;
        if (n == 0)
            break;

        std::cout << "Which permutation you want? ";
        int k;
        std::cin >> k;

        std::cout << GetKthPermutation(n, k) << std::endl;
    }
}

File: Go/KthPermutationSequence.go
package main

import (
	"fmt"
)

func GetPermutations(vec []byte, index int, permutations *[][]byte) {
	if index == len(vec) {
		c := make([]byte, len(vec))
		copy(c, vec)
		*permutations = append(*permutations, c)
		return
	}

	for i := index; i < len(vec); i++ {
		vec[i], vec[index] = vec[index], vec[i]
		GetPermutations(vec, index+1, permutations)
		vec[i], vec[index] = vec[index], vec[i]
	}
}

func GetKthPermutation(n, k int) string {
	vec := make([]byte, n)
	for i := 0; i < n; i++ {
		vec[i] = byte(i + 1)
	}

	var permutations [][]byte
	GetPermutations(vec, 0, &permutations)

	if len(permutations) >= k {
		r := permutations[k-1]
		for i := 0; i < len(r); i++ {
			r[i] += '0'
		}
		s := string(r)
		return s
	}
	return ""
}

func main() {
	for {
		fmt.Print("The number: ")
		var n, k int
		fmt.Scan(&n)
		if n == 0 {
			break
		}

		fmt.Print("Which permutation do you want? ")
		fmt.Scan(&k)

		v := GetKthPermutation(n, k)
		fmt.Println("The ", k, "th permutation of 1 to ", n, " is ", string(v))
	}
}
